File: go/events.go
package spiderbrowser

import "sync"

// EventHandler is a callback for spider-browser events.
type EventHandler func(data map[string]any)

// EventEmitter provides typed event pub/sub for spider-browser.
type EventEmitter struct {
	mu       sync.RWMutex
	handlers map[string][]EventHandler
	once     map[string][]EventHandler
}

// NewEventEmitter creates a new event emitter.
func NewEventEmitter() *EventEmitter {
	return &EventEmitter{
		handlers: make(map[string][]EventHandler),
		once:     make(map[string][]EventHandler),
	}
}

// On subscribes to an event.
func (e *EventEmitter) On(event string, handler EventHandler) {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.handlers[event] = append(e.handlers[event], handler)
}

// Off removes a handler for an event (removes the last matching one).
func (e *EventEmitter) Off(event string, handler EventHandler) {
	e.mu.Lock()
	defer e.mu.Unlock()
	// Remove by comparing function pointers isn't reliable in Go,
	// so we remove the last handler for simplicity.
	if handlers, ok := e.handlers[event]; ok && len(handlers) > 0 {
		e.handlers[event] = handlers[:len(handlers)-1]
	}
}

// Once subscribes to an event for a single firing.
func (e *EventEmitter) Once(event string, handler EventHandler) {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.once[event] = append(e.once[event], handler)
}

// Emit fires an event to all registered handlers.
func (e *EventEmitter) Emit(event string, data map[string]any) {
	e.mu.RLock()
	handlers := make([]EventHandler, len(e.handlers[event]))
	copy(handlers, e.handlers[event])
	onceHandlers := make([]EventHandler, len(e.once[event]))
	copy(onceHandlers, e.once[event])
	e.mu.RUnlock()

	for _, h := range handlers {
		h(data)
	}
	for _, h := range onceHandlers {
		h(data)
	}

	if len(onceHandlers) > 0 {
		e.mu.Lock()
		delete(e.once, event)
		e.mu.Unlock()
	}
}

// RemoveAllListeners removes all event handlers.
func (e *EventEmitter) RemoveAllListeners() {
	e.mu.Lock()
	defer e.mu.Unlock()
	e.handlers = make(map[string][]EventHandler)
	e.once = make(map[string][]EventHandler)
}

File: rust/src/events.rs
//! Lock-free event emitter for SpiderBrowser events.

use dashmap::DashMap;
use serde_json::Value;
use std::sync::Arc;

/// Callback type for event handlers.
pub type EventHandler = Arc<dyn Fn(Value) + Send + Sync>;

/// Browser type identifier.
pub type BrowserType = String;

/// Common browser type constants.
pub mod browsers {
    pub const AUTO: &str = "auto";
    pub const CHROME: &str = "chrome";
    pub const CHROME_NEW: &str = "chrome-new";
    pub const CHROME_H: &str = "chrome-h";
    pub const FIREFOX: &str = "firefox";
    pub const SERVO: &str = "servo";
    pub const LIGHTPANDA: &str = "lightpanda";
    /// Spider's custom Rust-based browser, optimized for fast scraping.
    pub const NAVI: &str = "navi";
}

/// Lock-free event emitter using DashMap for concurrent handler access.
#[derive(Clone, Default)]
pub struct SpiderEventEmitter {
    handlers: Arc<DashMap<String, Vec<EventHandler>>>,
}

impl SpiderEventEmitter {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn on(&self, event: &str, handler: EventHandler) {
        self.handlers
            .entry(event.to_string())
            .or_default()
            .push(handler);
    }

    pub fn off(&self, event: &str) {
        self.handlers.remove(event);
    }

    pub fn emit(&self, event: &str, data: Value) {
        if let Some(list) = self.handlers.get(event) {
            // Clone the vec to release the DashMap shard lock before invoking callbacks.
            let handlers = list.clone();
            drop(list);
            for handler in &handlers {
                let _ = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
                    handler(data.clone());
                }));
            }
        }
    }

    pub fn remove_all_listeners(&self) {
        self.handlers.clear();
    }
}
